### ISSUE TITLE
Improve ogg-mpg conversion

### ISSUE THREAD

**Issue by malrase** [wordset/wordset#128]
Right now, the sound-playback is very creepy/not well encoded.

**Comment by HamptonMakes** [comment]
I'm not sure if #127 would fix this on iOS 8/9

**Comment by HamptonMakes** [comment]
Currently, the audio sounds weird because we are using some weird Javascript library to try and get around the lack of OGG support. It's not very good though... we are using audio from Wikipedia and they only provide OGG versions, and the iPhone doesn't support OGG. So, we either need something that can decode OGG directly into the iOS waveforms, or the improvements in #127 might greatly increase the JS engine performance at doing it.

**Comment by rajkumariosdev** [comment]
I have added IDZAQAudioPlayer that provides ogg decoder/player. Currently, the ogg file is downloaded once we show the detail screen, I will update this to download only when user click the speaker button, for this I will also have to see how to show the loading indicator. There is a method evaluateJavascript in WKWebview, which I'm trying to use.

**Comment by rakeshak** [comment]
Downloading the ogg file after tapping the speaker icon would make the app feel slow; showing a loading indicator won't be of much help. If the current version works as below we shall leave it as it is:
1. Load the definitions.
2. After loading the definitions download the ogg file in the background (no loading indicator should be shown).
3. Play the ogg file immediately after a user taps the speaker icon.

Note:
1. This assumes that the user experience and performance of the app is more important than the cost of using the app. That is, if a user doesn't use the listen feature at all the bandwidth used for downloading the ogg file is wasted but it's a fair price to pay for the overall user experience of the app. 
2. If a user taps a synonym or the back button or any other link/button while the ogg file is downloading the app should immediately stop downloading and respond to the user's taps.

**Comment by HamptonMakes** [comment]
Yeah, I'd leave it as is right now... maybe we can make it an option that you can configure?

**Comment by rajkumariosdev** [comment]
I will update it to stop downloading when user taps any link in the page. Currently, it is downloading in background only and User will be free to click on any synonym/links.
@hcatlin I will for now add a flag that controls if the audio is downloaded immediately once detail screen is loaded or after user clicks defaulting to immediate download. You can change it later on if you wish.
The current implementation of immediate download is working fine without much delays.

**Comment by srkrambo** [comment]
This is done. I have made sure to cancel any previous audio downloads when we try to download new audio. I have also added a flag loadAudioImmediatelyOnLoad in Constants.swift which is true for now and ensures, we download the audio first and then show the play button. In case, you want to see how it behaves downloading after user presses play, you can set this to false.

**Comment by rakeshak** [comment]
In the latest build (7017) it seems the word is downloaded only after I tap the speaker icon—I'm shown a progress animation on the speaker icon. Hope this is taken care of in the final build.

For a particular session of the app usage, once downloaded, can the ogg file be saved in the cache so that if the user comes back to that word he can listen to it without waiting for it to download again? This might be too much work for a too small feature. Maybe this could be added to future enhancements.

**Comment by srkrambo** [comment]
The current implementation will not show any animation and we show the speaker icon only after the file is downloaded. It is possible to cache the files. May 4-5 hours of work maximum. @hcatlin Please confirm if we want it be done in 7.1 or future version. Also, the maximum cache size to be used, so that we don't use too user space.

**Comment by HamptonMakes** [comment]
Yeah, Rakesh the version you have is 7.0, this is all work for 7.1, so we don't have any beta builds yet until 7.0 hits the store.

The animation in 7.0 is downloading (streaming actually) and playing the audio.

Ideally, 7.1 should cache about 50mb, show the icon before the download is completed (but after we know there is a file), and use an animation during download/play.

Sound good?

—

**Comment by srkrambo** [comment]
Yes, it can be done. The only thing is, I need to figure out how to call the loading animation. There is a method evaluateJavaScript in WKWebview is particularly for that, but haven't used that before.

**Comment by rakeshak** [comment]
Yes, good; I was about to mention that we shouldn't wait to show the speaker icon till the download is complete. The speaker icon should be loaded along with the definitions, maybe immediately after loading the definitions and confirming that there is a sound file associated with the definition, as you said. 

Animation during download/play may not be required but would be nice to have. We'll design it for the next release, probably two—one for downloading and one for playing. In the meanwhile, we shall the remove the current animation and keep it static.

**Comment by HamptonMakes** [comment]
Rakesh: why not use the current one? Having no movement was really confusing if anything was responding. Mostly, people have their phone muted, etc, and it takes a second for the user to think "oh, I can't hear the sound" vs "the button does nothing".

Rajkumar, can you just put in start and stop hooks, and put a comment there for me. I'm really pretty good with the JS stuff, so I can do it tomorrow easily.

—

**Comment by srkrambo** [comment]
Sure will add comments where this need to be added.

**Comment by srkrambo** [comment]
Added TODO: in detailviewcontroller to start and end loading activity. I also updated loadAudioImmediatelyOnLoad to false. I will add caching mechanism tomorrow.

**Comment by rakeshak** [comment]
@hcatlin Sounds reasonable, we can use the current one as a stopgap. If possible in this version, we could try to implement the icon/link animation for touch feedback as in my latest prototype, it should solve the 'button does nothing' problem.

**Comment by srkrambo** [comment]
Implemented caching with cachesize 50 MB. The only item pending in this is the downloading animation.

**Comment by srkrambo** [comment]
Added loading indicator when downloading file. So the WKWebView stuff is done.
